src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+  private SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee ingestionDeliveryGuarantee =
+      SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE;
+
+  // ------ Streaming Ingest ------ //
+  // needs url, username. p8 key, role name
+  private final SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // Config set in JSON
+  private final Map<String, String> connectorConfig;
+
+  private final String taskId;
+
+  private final String streamingIngestClientName;
+
+  /**
+   * Key is formulated in {@link #partitionChannelKey(String, int)} }
+   *
+   * <p>value is the Streaming Ingest Channel implementation (Wrapped around TopicPartitionChannel)
+   */
+  private final Map<String, TopicPartitionChannel> partitionsToChannel;
+
+  public SnowflakeSinkServiceV2(
+      SnowflakeConnectionService conn, Map<String, String> connectorConfig) {
+    if (conn == null || conn.isClosed()) {
+      throw SnowflakeErrors.ERROR_5010.getException();
+    }
+
+    this.fileSize = SnowflakeSinkConnectorConfig.BUFFER_SIZE_BYTES_DEFAULT;
+    this.recordNum = SnowflakeSinkConnectorConfig.BUFFER_COUNT_RECORDS_DEFAULT;
+    this.flushTime = SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_DEFAULT;
+    this.conn = conn;
+    this.recordService = new RecordService();
+    this.telemetryService = conn.getTelemetryClient();
+    this.topic2TableMap = new HashMap<>();
+
+    // Setting the default value in constructor
+    // meaning it will not ignore the null values (Tombstone records wont be ignored/filtered)
+    this.behaviorOnNullValues = SnowflakeSinkConnectorConfig.BehaviorOnNullValues.DEFAULT;
+
+    this.connectorConfig = connectorConfig;
+    // Streaming ingest requires a properties.
+    // TODO request a builder pattern for SI Client.
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(connectorConfig);
+    this.taskId = connectorConfig.getOrDefault(Utils.TASK_ID, "-1");
+    this.streamingIngestClientName =
+        STREAMING_CLIENT_PREFIX_NAME + conn.getConnectorName() + "_" + taskId;
+    this.streamingIngestClient =
+        SnowflakeStreamingIngestClientFactory.builder(this.streamingIngestClientName)
+            .setProperties(streamingClientProps)
+            .build();
+    this.partitionsToChannel = new HashMap<>();
+  }
+
+  @Override
+  public void startTask(String tableName, String topic, int partition) {
+    String partitionChannelKey = partitionChannelKey(topic, partition);
+    LOGGER.info(
+        "Opening a channel with name:{} for table name:{}, topic:{}, partition:{}",
+        partitionChannelKey,
+        tableName,
+        topic,
+        partition);
+    // the table should be present before opening a channel so lets do a table existence check here
+    createTableIfNotExists(tableName);
+    SnowflakeStreamingIngestChannel partitionChannel =
+        streamingIngestClient.openChannel(getOpenChannelRequest(partitionChannelKey, tableName));
```

Comment:
What if the channel has been used before? Do we need to start from the last offset token?

Reply:
Hmm good question. So only during first init (open) https://kafka.apache.org/24/javadoc/org/apache/kafka/connect/sink/SinkTask.html we wont need the offset token. 

But lets say there is a rebalance, do we need to pass in offsetToken in openChannel? Also, the partitionChannelKey is still the same since the key is same before and after rebalance. 

I thought we would fetch offsetToken only when we get records from kafka and when we need to reset the offsets based on what we get from snowflake (getOffsetToken API)

Reply:
Also looking at code, offsetToken is fetched from an internal call. So I am assuming from the client side it is automatically handled if the same name is used?

Reply:
yes, you're correct, we would fetch the offset token as part of the openChannel, so I assume you need to somehow reset the offset in Kafka after reopening each channel?